pkg/traces/exporter/exporter.go

```diff
@@ -0,0 +1,138 @@
+package exporter
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/kolide/kit/version"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/osquery"
+	"github.com/kolide/launcher/pkg/traces"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+)
+
+var archAttributeMap = map[string]attribute.KeyValue{
+	"amd64": semconv.HostArchAMD64,
+	"386":   semconv.HostArchX86,
+	"arm64": semconv.HostArchARM64,
+	"arm":   semconv.HostArchARM32,
+}
+
+type TraceExporter struct {
+	provider *trace.TracerProvider
+}
+
+// NewTraceExporter sets up our traces to be exported via OTLP over HTTP.
+// On interrupt, the provider will be shut down.
+func NewTraceExporter(ctx context.Context, serverProvidedDataStore types.Getter, client osquery.Querier) (*TraceExporter, error) {
+	traceClient := otlptracehttp.NewClient(otlptracehttp.WithInsecure())
+	exp, err := otlptrace.New(ctx, traceClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not create exporter for traces: %w", err)
+	}
+
+	traceExporter := &TraceExporter{}
+
+	// Set up the exporter in the background -- we need to wait for the osquery client to be available, but don't
+	// want to block on it.
+	go traceExporter.configureProvider(exp, serverProvidedDataStore, client)
+
+	return traceExporter, nil
+}
+
+// configureProvider sets up the trace provider attributes describing this application and installation,
+// uniquely identifying the source of these traces.
+func (t *TraceExporter) configureProvider(exp *otlptrace.Exporter, serverProvidedDataStore types.Getter, client osquery.Querier) {
+	// Set all the attributes that we know we can get first
+	attrs := []attribute.KeyValue{
```

Comment:
Is `attrs` something that we can change in runtime? Because I think it's going to be hard to populate some of this early in startup. `client osquery.Querier`, for example, won't be online as early as we want this.

But if we can adjust them after we're running, we get nice behavior. (And CJ has a PR with a callback mechanism somewhere)

Reply:
I can take another look, but I didn't see support for updating the provider after it's already been set. The way I have this working now is that the exporter just waits till the osquery client is ready, so the first ~30sec-1min of traces wouldn't be reported -- which would be a bummer for e.g. trying to figure out osquery extension startup behavior via traces.

I think we could replace the provider, so we could do something like initially writing traces to a file until the osquery client is ready, and then starting to export traces to an external collector once we can get the identifiable information we need.

I'll research more.

Reply:
So, unfortunately, it looks like the provider's resource (i.e. all these identifying attributes about the device) is expected to be immutable once created. I did get it working by creating an initial provider with the info we had and replacing it with a new provider with more attributes once we could access the osquery client -- but then it would refuse to ship traces if I called `Shutdown` on the old provider, which we are required to do to free up resources. Trying something different next.

Reply:
Never mind! This approach actually worked great once I made sure to use a new exporter for the new provider, since the old exporter was getting shut down.